libjxl: Fix tests on darwin

jdelStrother opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

On my intel macbook running 12.0.1, building libjxl fails while running its LibraryCLinkageTest - 

```
1122/1129 Test #1128: LibraryCLinkageTest .......................................................................................................Subprocess aborted***Exception:   0.32 sec
dyld[53579]: Library not loaded: @rpath/libjxl.0.dylib
  Referenced from: /private/tmp/nix-build-libjxl-0.5.drv-1/source/build/tools/tests/libjxl_test
  Reason: tried: '/nix/store/82d092jfk2sckailcp5hpb29xsxhk4h2-swift-corefoundation/Library/Frameworks/libjxl.0.dylib' (no such file), '/nix/store/82d092jfk2sckailcp5hpb29xsxhk4h2-swift-corefoundation/Library/Frameworks/libjxl.0.dylib' (no
```

I _think_ this is an appropriate fix, and does solve the problem for my local builds. However, it seems like hydra is already [successfully building the job](https://hydra.nixos.org/job/nixpkgs/nixpkgs-21.11-darwin/libjxl.x86_64-darwin) without this fix - I'm not sure how. Anyone know if it's due to differences in macos versions or something else?


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

jdelStrother committed:
libjxl: Fix tests on darwin

This fixes the LibraryCLinkageTest, which was failing to find the built library:

 1122/1129 Test #1128: LibraryCLinkageTest .......................................................................................................Subprocess aborted***Exception:   0.32 sec
dyld[53579]: Library not loaded: @rpath/libjxl.0.dylib
  Referenced from: /private/tmp/nix-build-libjxl-0.5.drv-1/source/build/tools/tests/libjxl_test
  Reason: tried: '/nix/store/82d092jfk2sckailcp5hpb29xsxhk4h2-swift-corefoundation/Library/Frameworks/libjxl.0.dylib' (no such file), '/nix/store/82d092jfk2sckailcp5hpb29xsxhk4h2-swift-corefoundation/Library/Frameworks/libjxl.0.dylib' (no

nh2 commented:
> However, it seems like hydra is already [successfully building the job](https://hydra.nixos.org/job/nixpkgs/nixpkgs-21.11-darwin/libjxl.x86_64-darwin) without this fix - I'm not sure how. Anyone know if it's due to differences in macos versions or something else?

Unfortunately, I don't know enough about macOS to tell, somebody else will have to answer that question and whether this change is needed/correct.

risicle commented:
Do you have the sandbox enabled?

risicle reviewed (approved):
Looks sensible to me, builds, macos 10.15.

Quite possible different macos versions behave differently here.

jdelStrother commented:
> Do you have the sandbox enabled?

No, should I?

risicle commented:
Not particularly, I was just trying to figure out what was "different", but it seems likely os version is to blame.

risicle commented:
Oh I've just noticed this is aimed at master and staging already has libjxl 0.6.1. It would probably be best to retarget this at staging and retest.

jdelStrother commented:
I'll try staging. It's been building on my macbook for a few hours now though - looks like it'll take a while longer...

Is there any guidance anywhere which branch to submit PRs against? It's not mentioned in, eg, CONTRIBUTING.md

risicle commented:
Extremely broad rule of thumb: if it generates more than 500 rebuilds, send it to staging. In this case though it's accidentally intertwined with something that's already in staging - that bump went to staging because it was generated by @r-ryantm , a bot, which is a bit conservative and will aim PRs with 101-500 rebuilds to staging too.

risicle commented:
p.s. rebuilds on staging are quite taxing. Often *I* will cherry-pick the relevant commit(s) back to master for testing and deem that "good enough". In this case I would cherry-pick both the 0.6.1 bump and then your commit and test that.

jdelStrother commented:
> Oh I've just noticed this is aimed at master and staging already has libjxl 0.6.1. It would probably be best to retarget this at staging and retest.

OK, I've rebased on staging and resubmitted. (The same rpath error was present in the newer version, and fixed by this PR)

> if it generates more than 500 rebuilds, send it to staging
Is there a way to see how many packages are going to need rebuilding, following a given package change?

> p.s. rebuilds on staging are quite taxing
no kidding, it occupied my MBP for a good 16 hours or so 😬

jdelStrother commented:
(Out of academic curiosity I tried reproducing the rpath error on EC2 mac instances. 10.15.7 & 11.6.1 build fine, it looks like the problem is new in 12.0.1)

ryantm commented:
> Is there a way to see how many packages are going to need rebuilding, following a given package change?

https://github.com/NixOS/ofborg/blob/released/ofborg/src/outpaths.nix is the Nix code OfBorg uses for this.

risicle commented:
As a very quick proxy I tend to just look at the previous PR for a package and see what ofborg thought about that. Otherwise I'll do a `nixpkgs-review` and stop it once it starts building.

nh2 commented:
> Is there any guidance anywhere which branch to submit PRs against? It's not mentioned in, eg, CONTRIBUTING.md

It is in the nixpkgs manual, which is linked from `CONTRIBUTING.md`, but the link was not very prominent.

I made a PR to improve it here: https://github.com/NixOS/nixpkgs/pull/153513
